Full Dubins Airplane implementation

rafael1193 opened the pull request:
The Dubins Airplane (a.k.a. dubins3d) implementation should be fully usable. The missing part, getting the trajectory drawn, is complete for every kind of trajectory.

New features at a glance:
 - __Dubins3dPathLength__ has a new method __sample__ that returns the position and direction of the uav for a given t∈[0, L_2d). It works pretty much the same as the original dubins2d counterpart.
 - the __UAV__ class __path_sampling__ method is exposed in the python interface.
 - python/fire_rs/planning/demo_dubins3dpath.py showcases airplane paths for every altitude goal.

The Dubins Airplane code works hassle when using it with the sole purpose of finding the pah between two points. However, issues appear when using the code while planning (See bug #9). python/fire_rs/planning/test_dubins3dpath.py contains some waypoint sequences found failing during planing, but not in this script. The most common point of failure is not being able to find a solution in _helix_optimization_.

rafael1193 committed:
No opt<> in Dubins3dPathLength

rafael1193 committed:
make helix_optimization const

rafael1193 committed:
Fix tests

rafael1193 committed:
Comment Dubins3dPath

rafael1193 committed:
Implement a sampling function for Dubins 3d paths

Medium altitude paths are not ready yet

rafael1193 committed:
Use the Dubins3d path sampler for 3d paths

rafael1193 committed:
Expose UAV.path_sampling in python

rafael1193 committed:
Add Dubins3d path demo script

It helps identifying regressions in the sampling code

rafael1193 committed:
Implement the medium altitude case

rafael1193 committed:
Showcase low- medium- and high-altitude paths

rafael1193 committed:
Some corrections and cleanup

rafael1193 committed:
Expose dubins_init_with_type in dubins.h

rafael1193 committed:
Test script for some failing Dubins 3d paths while planning

arthur-bit-monnot commented:
A bit complex for me to really dive in right now.

The fact that problems only occur while planning is probably simply to planning using the method **a lot** and on many different outputs.
You should probably try to isolate which input is problematic, for instance by inspecting the call that makes the assert failure (should be easily doable with a break point in helix optimization). 
Another option is to generate many random test cases until it crashes (this a good thing to do in general as it tests the logic of the function independently from the rest of the code).
Once we have an input that produces invalid results, it will be much easier to debug =)

rafael1193 commented:
Well, that's exactly what I did; I took the problematic call out. 

The interesting thing is that it works when the call is isolated. It looks like we found a _"heisenbug"_ 😮

rafael1193 committed:
Merge branch 'master'

arthur-bit-monnot commented:
I would say that the python tests are to far away from the actual crash.
Try to call helix_optimization directly with the same arguments than those that provoked the crash.

If the function does not crash with exactly the same arguments it means it is not pure, i.e., (i) it accesses data outside of the arguments (likely) or (ii) its arguments are modified while its executing.
For (i) it means that the function is either affected by some state (possible since it is in an object) or because it does an erroneous memory access (dangling pointer or access out of array bounds).
(ii) is possible in case of multithreading; erroneous memory accesses might also cause similar behaviour.

Hence my advice is : make reeeally sure that the bug is not inside the function. Otherwise I guess we should start debugging memory accesses. The most practical tool for this is Valgrind (very efficient but also very annoying to use).

rafael1193 committed:
Stop the helix_optimization if the convergence stalls

Closes #9

rafael1193 committed:
Stop the spiral_extension loop if the convergence process stalls

Apply the same strategy as used in helix_optimization.

rafael1193 committed:
Replace Dubins3d MAX_LOOPS assertions by warnings

arbimo commented on cpp/src/core/structures/trajectory.h:
```diff
@@ -213,7 +214,9 @@ class Trajectory {
         auto increase_in_length =  insertion_length_cost(insert_loc, segment);
-        ASSERT(ALMOST_GREATER_EQUAL(increase_in_length, 0))
+        if (!ALMOST_GREATER_EQUAL(increase_in_length, 0)) {
+            std::cerr << "WARNING: increase in length < 0 : " << increase_in_length << std::endl;
```
In general, I'd rather have an assert because it forces us to handle the problem when it shows up and it makes the problem appear in automated tests or large runs.

Here failure to fullfill the test would not show up in continuous integration. If the assert is to strict, it might be relaxed to make sure reasonable results are accepted.

rafael1193 replied:
You are right. I dropped the commit.

rafael1193 committed:
Separate Dubins3dPath and Dubins3dPathLength

Recover this separation. Dubins3dPathLength only calculates the expected
length of the 3d path (which is quite fast to compute). Then, the rest
of the geometrical computation is in the inherited class Dubins3dPath,
where the geometrical math is done.

There should not be more problems with the "increase cost" assertions
failing because Dubins3dPathLength returns exact path length.

rafael1193 committed:
Merge branch 'master' into fulldubins3d

rafael1193 commented:
The branch should be ready to be merged in master. No errors so far in travis, nor the unittest, nor manual launch of benchmark.py.

@arthur-bit-monnot can you make a last test to see if it is ok on your side?

rafael1193 commented:
I ran thousands of random trajectories: no errors, no warnings. Merging

arthur-bit-monnot commented:
@rafael1193 
Good ! Consider adding adding the tests to the travis configuration so they are run automatically.
This could be benchmark.py without the elevation=flat option and common C++ unit tests.
(right now I think the C++ tests still fail to build due to missing boost test dependency in docker).
